enhancements/etcd/automated-backups-no-config.md

```diff
@@ -0,0 +1,205 @@
+
+
+## Proposal
+- Provide a default etcd backup configuration as a `EtcdBackup` CR.
+- Provide a default reliable storage mechanism to save the backup data on.
+- Provide a separate feature gate for the automated backup with default config.
+
+
+### Workflow Description
+- The user will enable the AutomatedBackupNoConfig feature gate.
+- A `EtcdBackup` CR is being installed on the cluster.
+- A PVC and PV are being created to save the backup data reliably.
+
+
+### API Extensions
+- No API changes are required. 
+- See https://github.com/openshift/enhancements/blob/master/enhancements/etcd/automated-backups.md#api-extensions
+
+### Topology Considerations
+TBD
+#### Hypershift / Hosted Control Planes
+TBD
+#### Standalone Clusters
+TBD
+#### Single-node Deployments or MicroShift
+TBD
+
+### Implementation Details/Notes/Constraints [optional]
+- Need to agree on a default schedule.
+- Need to agree on a default retention policy.
+
+- Several options exist for the default `PVCName`.
+  - Relying on `dynamic provisioning` is sufficient, however not an option for `SNO` or `BM` clusters.
+  - Utilising `local storage operator` is a proper solution, however installing a whole operator is too much overhead.
+  - The most viable solution to cover all OCP variants is to use `local volume`.
+    Please find below this solution's Pros & Cons.
+    - Pros :- 
+      - The PV will be bound to a specific master node.
+      - The Backup Pod will be scheduled always to this master node since it uses a PVC bound to this PV.
+      - Using an SSD is possible, since local volume allows mounting it into a Pod.
+      - Having the Backup Pod scheduling deterministic is vital, since the retention policy needs the recent backup in place to work.
+    - Cons :-
+      - If the master node where the PV is mounted became unhealthy/unavailable/unreachable etc.
+      - The backups are no longer accessible, also taking new backups is no longer possible.
+      - Ideally, the backups should be taken on a round-robin fashion on each master node to avoid overwhelming a specific etcd member more than others.
+      - Also spreading the backups among all etcd cluster members provide guarantees for disaster recovery in case of losing two members at the same time.
+      - However, using the `local volume` option over all the master nodes will be complicated and error prune.
+
+### Risks and Mitigations
+
+When the backups are configured to be saved to a `local` type PV, the backups are all saved to a single master node where the PV is provisioned on the local disk.
+
+In the event of the node becoming inaccessible or unscheduled, the recurring backups would not be scheduled. The periodic backup config would have to be recreated or updated with a different PVC that allows for a new PV to be provisioned on a node that is healthy.
+
+If we were to use `localVolume`, we need to find a solution for balancing the backups across the master nodes, some ideas could be
```

Comment:
how would we restore from a localVolume? Is the data stored somewhere in `/var/`?

Reply:
so the localVolume is being mounted anywhere as specified in the Pod. In fact, you can mount a whole external drive. 

iiuc, `hostPath` supports only mount points from the node's FS

Reply:
imagine we have to run the restore script, how would we get the local volume mounted into a recovery machine? would we create another static pod that would mount the localVolume?

I imagine the local storage plugin must be running along with the kubelet for that to work? @jsafrane / @gnufied maybe you guys can elaborate a bit how that works, I don't have much time this week to read through the CSI implementation :racehorse:

Reply:
pardon my ignorance for not being familiar with etcd deployment, but even if there is a default SC available, why are we backing up just one replica of etcd? I thought, we would want to backup all replicas no?  Such as - what if in case of network partition or something, the replica we are backing up is behind?


So IMO it sounds whether we are using LocalVolume or something else, we should always be creating backups across master nodes? Is that accurate?

Reply:
You can also use a directory as a `localvolume` if statically provisioning a local PV. So basically `path: ` field points to a directory. The difference with `hostPath` volumes is, the container using local-volume doesn't have to run with elevated privileges.

Reply:
> I thought, we would want to backup all replicas - no? Such as - what if in case of network partition or something, the replica we are backing up is behind?

Snapshots are always generated from the leader, so it's never behind. We only want to _store_ the snapshot on all nodes. 

>  Local Volumes are a inbuilt feature of k8s and hence no additional plugin is necessary

that's great to hear, thanks for the docs :) 

> the container using local-volume doesn't have to run with elevated privileges.

the restore script/ pod will have to run with elevated permissions anyway, because of static pods and host networking. We don't have that permission constraint at all, which is why hostPaths are somewhat convenient.

Reply:
@tjungblu 

> imagine we have to run the restore script, how would we get the local volume mounted into a recovery machine? would we create another static pod that would mount the localVolume?

So the localVolume will be created with `Retain` policy. This way the `PV` will still have the same files and can be mounted in the new etcd pod and the backup will exist. 

However, if the whole machine is unhealthy/unreachable, then the backup will not be accessible.

Reply:
> ... and can be mounted in the new etcd pod and the backup will exist.

check the restore script, you will need to unpack the snapshot before you can run any pod

Reply:
@gnufied in order to make sure the backups are accessible, and that we can round-robin the backups across all master nodes, which are up-to-date with the `etcd-cluster` leader

Is it possible to 

- Create StatefulSet across the master nodes, where the PVC template uses the `localVolume` to provision PV.
- Since we need to do the backup according to a schedule as per [EtcdBackupSpec](https://github.com/openshift/api/blob/0689f006bcdeccf685c75438835beffa8bb14db7/config/v1alpha1/types_backup.go#L46), the issue to manage the backup pods within the STS using the CronJob
- I am not aware if this is possible, but at least, we can generate `Event` from the CronJob and the STS could react to these event by taking a backup

Reply:
> check the restore script, you will need to unpack the snapshot before you can run any pod

We can use `init-container` to unpack the backup, and then the etcd-pod can start ? 

Otherwise, we can start a fresh etcd-pod and then run `etcdctl restore` ? 

Please correct me if I am wrong

Reply:
> * Create StatefulSet across the master nodes, where the PVC template uses the `localVolume` to provision PV.

I believe you mean DaemonSet across the master nodes :wink:

Reply:
I actually thought about this as well, but there is no `PersistentVolumeClaim` template on DaemonSet.

Having a separate PVC & PV for each master node would be ideal for spreading the backups across the master nodes. 

That being said, according to the latest update, we are going to use `SideCar` container && `hostPath` volume